app/questionnaire/navigator.py

```diff
@@ -1,193 +1,288 @@
 
-        can_reach_summary = self.can_reach_summary(answers, routing_path)
+        can_reach_summary = self.can_reach_summary(routing_path)
 
         # Make sure we don't update original
-        location_path = list(Navigator.PRECEEDING_INTERSTITIAL_PATH)
-        location_path.extend(routing_path)
+        location_path = [{
+            "block_id": block_id,
+            "group_id": self.first_group_id,
+            "group_instance": 0,
+        } for block_id in Navigator.PRECEEDING_INTERSTITIAL_PATH]
+
+        location_path += routing_path
 
         if can_reach_summary:
-            location_path.extend(Navigator.CLOSING_INTERSTITIAL_PATH)
+            for block_id in Navigator.CLOSING_INTERSTITIAL_PATH:
+                location_path.append({
+                    "block_id": block_id,
+                    "group_id": self.last_group_id,
+                    "group_instance": 0,
+                })
 
         return location_path
 
+    def get_first_group_id(self):
+        return self.survey_json['groups'][0]['id']
+
+    def get_last_group_id(self):
+        return self.survey_json['groups'][-1]['id']
+
     def get_first_block_id(self):
         return self.survey_json['groups'][0]['blocks'][0]['id']
 
     def get_blocks(self):
         blocks = []
         for group in self.survey_json['groups']:
-            blocks.extend([block for block in group['blocks']])
+            blocks.extend([{
+                "group_id": group['id'],
+                "group_instance": 0,
+                "block": block,
+            } for block in group['blocks']])
+
+            if 'routing_rules' in group:
+                for rule in group['routing_rules']:
+                    if 'repeat' in rule.keys():
+                        no_of_times = evaluate_repeat(rule['repeat'], self.answer_store)
+                        for i in range(1, no_of_times):
+                            blocks.extend([{
+                                "group_id": group['id'],
+                                "group_instance": i,
+                                "block": block,
+                            } for block in group['blocks']])
         return blocks
 
-    @staticmethod
-    def get_first_location():
-        return Navigator.PRECEEDING_INTERSTITIAL_PATH[0]
+    @classmethod
+    def get_first_location(cls):
+        return cls.PRECEEDING_INTERSTITIAL_PATH[0]
 
-    def get_next_location(self, answers=None, current_location_id=None):
+    def _get_current_location_index(self, current_group_id, current_block_id, current_iteration):
+        current_group_id = current_group_id or self.first_group_id
+
+        this_block = {
+            "block_id": current_block_id,
+            "group_id": current_group_id,
+            "group_instance": current_iteration,
+        }
+
+        if this_block in self.location_path:
+            return self.location_path.index(this_block)
+        return -1
```

Comment:
Hah, yeah, that's what I meant 😛  I think returning `None` says 'there is no current location index' more clearly than `-1`